### ISSUE TITLE
Incorrect Visualization

### ISSUE THREAD

**Issue by Jayden9912** [NVlabs/SegFormer#117]
![image](https://user-images.githubusercontent.com/85933053/217203082-20c644a2-8d20-41ca-9fb5-49f429980cc2.png)
I think the order of overlapped patch merging is wrong. Is my comment correct?

**Comment by AlexandreBrown** [comment]
Hi,  
I agree with you and this was the confusing part of the paper for me.  
From my understanding, the Overlap Patch Embedding is obtained by doing `Conv2d` on the input image with the appropriate patch size, stride and padding such that we get output shape `(C_1, H//4, W//4)` for the first stage, then `(C_2, H//8, W//8)` for the second stage, `(C_3, H//16, W//16)` for the third stage and `(C_4, H//32, W//32)` for the fourth stage (here H = original image height and W = original image width ex: if image is `3x512x512` then `H=512`, `W=512`).   
The parameters described in the `Overlapped Patch Merging` section of the paper are exactly the right parameters to use for each stage in order to obtained the embeddings of the right dimensions.  
![image](https://github.com/NVlabs/SegFormer/assets/26939775/bdf03a9d-1380-42a0-be20-3b57c1e58046)
Ex:  
Given an Input image of shape : `Bx3x512x512` (format `BxCxHxW` used here where B = batch size, C = input channel, H = height and W = width)  
Let's use `c_1 = 64` (Depends on if you want to use b0 or b1-b5 variant)  
![image](https://github.com/NVlabs/SegFormer/assets/26939775/f202d4e3-84ca-45a0-b144-d6010056e2e2)
  
Then the convolution to obtain the Overlap Patch Embeddings for **stage 1** would be :  
```python
K = 7
S = 4
P = 3
nn.Conv2d(in_channels=3, 
          out_channels=c_1, 
          kernel_size=K, 
          stride=S, 
          padding=(P,P))
```

The embeddings output shape obtained from this conv would be:  `Bx64x128x128` (`64 = c_1`).  
Note how `128` is exactly `512 // 4`.    

Same principle applies for the other stages  but for stage 2 we will get `512//8`, stage 3: `512//16` and stage 4: `512//32`.  
  
Therefore it seems to me like the section `Overlapped Patch Merging` is describing how to get the overlapping patch embeddings and such step seems to be performed **before** passing the input to the transformer block.  
With this information I would agree with the claim that perhaps the visualization is wrong? Or did I miss something ?

**Comment by AlexandreBrown** [comment]
I agree, your visualization seems like a more accurate representation of the model.
